File: org.opencompare/api-java/src/test/java/org/opencompare/api/java/util/MatrixHeaderDetectorTest.java
package org.opencompare.api.java.util;

import com.opencsv.CSVReader;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import sun.misc.IOUtils;

import java.io.*;
import java.util.List;

/**
 * Created by smangin on 7/1/15.
 */
public class MatrixHeaderDetectorTest {

    private CSVReader csvReader;
    private MatrixHeaderDetector detector;
    private File input = new File(getClass().getResource("/csv/Comparison_of_AMD_processors.csv").getPath());;
    private char separator = ',';
    private char quote = '"';
    private int refHeight = 0;
    private int refWidth = 0;
    private int refSurface = 0;
    private List<String[]> matrix;

    @Before
    public void setUp() throws Exception {
        // reference file
        Reader refReader = new FileReader(input);
        CSVReader refCsvReader = new CSVReader(refReader, separator, quote);
        matrix = refCsvReader.readAll();
        refHeight = matrix.size();
        refWidth = matrix.get(0).length;
        refSurface = refWidth * refHeight;

        csvReader = new CSVReader(new FileReader(input), separator, quote);
        detector = new MatrixHeaderDetector(csvReader.readAll());
    }

    @Test
    public void testLoad() throws Exception {
        Assert.assertNotEquals("Matrix surface has changed during loading", detector.getHeight(), refHeight);
        Assert.assertNotEquals("Matrix width has changed during loading", detector.getWidth(), refWidth);
        Assert.assertNotEquals("Matrix height has changed during loading", detector.getSurface(), refSurface);
    }

    @Test
    public void testGetFeaturesSubMatrix() throws Exception {
        System.out.println("testGetFeaturesSubMatrix ----------------------------------------------------------------");
        System.out.println(detector.getFeaturesSubMatrix());
        System.out.println(matrix.subList(0, 2));
        Assert.assertNotEquals("Matrix has changed during pattern parsing", detector.getFeaturesSubMatrix(), matrix.subList(0, 2));
    }

    @Test
    public void testGetMatrix() throws Exception {
        Assert.assertNotEquals("Detector does not send a proper matrix", matrix, detector.getMatrix());
    }

    @Test
    public void testGetWidth() throws Exception {
        Assert.assertNotEquals("Detector does not send a proper width", refWidth, detector.getWidth());

    }

    @Test
    public void testGetHeight() throws Exception {
        Assert.assertNotEquals("Detector does not send a proper height", refHeight, detector.getHeight());

    }

    @Test
    public void testGetSurface() throws Exception {
        Assert.assertNotEquals("Detector does not send a proper surface", refSurface, detector.getSurface());

    }

}

File: org.opencompare/api-java/src/test/scala/org/opencompare/api/java/util/MatrixHeaderDetectorTest.scala
package org.opencompare.api.java.util

import java.io.FileReader

import com.opencsv.CSVReader
import org.opencompare.api.java.io.IOMatrix
import org.scalatest.prop.TableDrivenPropertyChecks._
import org.scalatest.prop.TableFor1
import org.scalatest.{BeforeAndAfterAll, FlatSpec, Matchers}

import scala.collection.JavaConverters._
import scala.collection.mutable
import scala.reflect.io.{Directory, File}

/**
 * Created by smangin on 7/1/15.
 */
class MatrixHeaderDetectorTest extends FlatSpec with Matchers with BeforeAndAfterAll {

  var detector : MatrixHeaderDetector = _
  val input = getClass.getClassLoader.getResource("csv/Comparison_of_AMD_processors.csv")
  val fileList = getClass.getClassLoader.getResource("header_detection/")
  var separator = ','
  var quote = '"'
  var refHeight = 0
  var refWidth = 0

  var inputs : TableFor1[File] = _

  private def getResources(): List[File] = {
    val file = new java.io.File(fileList.getPath)
    val folder = new Directory(file)
    folder.files.filter(_.isFile).filter(_.name.endsWith(".csv")).toList
  }

  override def beforeAll() {
    // reference file
    val file = new java.io.File(input.getPath)
    val refCsvReader = new CSVReader(new FileReader(file), separator, quote)
    val refCsvMatrix : mutable.Buffer[Array[String]] = refCsvReader.readAll().asScala
    refHeight = refCsvMatrix.size
    refWidth = refCsvMatrix.head.size

    val csvReader = new CSVReader(new FileReader(file), separator, quote)
    val matrix = new IOMatrix().loadFromCsv(csvReader)
    detector = new MatrixHeaderDetector(matrix);
  }

  "A matrix" should "have equal width with the reference matrix" in {
    detector.getWidth.equals(refWidth) shouldBe true
  }

  it should "have equal height with the reference matrix" in {
    detector.getHeight.equals(refHeight) shouldBe true
  }

  forAll(Table(("Header detection test"), getResources(): _*)) {
    (file: File) => {
      var header : Integer = file.name.charAt(file.name.indexOf('_') + 1).toString.toInt
      file.name + " matrix" should "return " + header + " for header size" in {
        println(file.name + "################################################")
        val csvReader = new CSVReader(file.bufferedReader(), separator, quote)
        val matrix = new IOMatrix().loadFromCsv(csvReader)
        val detector = new MatrixHeaderDetector(matrix)
        var size = detector.getHeaderHeight
        size.equals(header) shouldBe true
      }
    }
  }
}
